tools/dataflow-throttling/src/main/java/com/google/cloud/pso/dataflowthrottling/DynamicThrottlingTransform.java

```diff
@@ -0,0 +1,332 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(ThrottlingOrchestration.class);
+
+    /**
+     * Define tuple tags to process corresponding payload response.
+     */
+    private final TupleTag<OutputT>  successTag = new TupleTag<OutputT>() {};
+    private final TupleTag<Result<InputT>> errorTag = new TupleTag<Result<InputT>>() {};
+    private final TupleTag<Result<InputT>> throttlingTag = new TupleTag<Result<InputT>>() {};
+
+    private final int numberOfGroups;
+    private final java.time.Duration batchInterval;
+    private final java.time.Duration resetCounterInterval;
+    private final int numOfEventsToBeProcessedForBatch;
+    private final int kInRejectionProbability;
+    private final Coder<InputT> elemCoder;
+    private final ClientCall<InputT, OutputT> clientCall;
+
+    /**
+     * Builder class to set either properties defined by client or to default.
+     * @param <BuilderInputT> Type of the event from the source.
+     * @param <BuilderOutpuT> Type of the enriched event.
+     */
+    public static class Builder<BuilderInputT, BuilderOutpuT>{
+        private int kInRejectionProbability=1;
+        private java.time.Duration resetCounterInterval= java.time.Duration.ofMinutes(1);
+        private java.time.Duration batchInterval= java.time.Duration.ofSeconds(1);
+        private int numberOfGroups=1;
+        private int numOfEventsToBeProcessedForBatch;
+        private Coder<BuilderInputT> elemCoder;
+        private ClientCall<BuilderInputT, BuilderOutpuT> clientCall;
+
+        public Builder(ClientCall<BuilderInputT, BuilderOutpuT> clientCall, Coder<BuilderInputT> elemCoder, int numOfEventsToBeProcessedForBatch) {
+            this.clientCall = clientCall;
+            this.elemCoder = elemCoder;
+            this.numOfEventsToBeProcessedForBatch = numOfEventsToBeProcessedForBatch;
+        }
+
+        public Builder<BuilderInputT, BuilderOutpuT> withResetCounterInterval(java.time.Duration resetCounterInterval){
+            this.resetCounterInterval = resetCounterInterval;
+            return this;
+        }
+
+        public Builder<BuilderInputT, BuilderOutpuT> withBatchInterval(java.time.Duration batchInterval){
+            this.batchInterval = batchInterval;
+            return this;
+        }
+
+        public Builder<BuilderInputT, BuilderOutpuT> withNumberOfGroups(int numberOfGroups){
+            this.numberOfGroups = numberOfGroups;
+            return this;
+        }
+
+        public Builder<BuilderInputT, BuilderOutpuT> withKInRejectionProbability(int kInRejectionProbability){
+            this.kInRejectionProbability=kInRejectionProbability;
+            return this;
+        }
+
+        public DynamicThrottlingTransform<BuilderInputT, BuilderOutpuT> build(){
+            return new DynamicThrottlingTransform<BuilderInputT, BuilderOutpuT>(this);
+        }
+    }
+
+    /**
+     * Result object that should pass to the throttlingTag and errorTag.
```

Comment:
* Result object that will be passed to the throttling and error output pcollection.